validate.py

```diff
@@ -148,29 +170,31 @@ def generate_result_text(query_name, checked_query_output_df):
     job_name = job["full_name"]
     query_keys = job["queries"]
 
-    # Set up connections to data sources used by job's queries
-    manager = DBConnManager()
-    manager.set_up(query_keys)
+    data_sources_for_queries = [QUERIES[query_key]['data_source'] for query_key in query_keys]
+    data_sources_used_by_job = set(data_sources_for_queries)
+
+    data_source_data: dict[DataSourceName, DataSource] = {
+        k: v for k, v in ENV['DATA_SOURCES'].items() if k in data_sources_used_by_job
+    }
+    db_manager = DBConnManager(data_source_data)
 
     # Execute queries and generate results text
-    results_text = ""
-    flags = []
-    for query_key in query_keys:
-        query = QUERIES[query_key]
-        query_output_df = execute_query_and_write_to_csv(query, manager)
-        checks_result = run_checks_on_output(query['checks'], query_output_df)
-        flags += checks_result.flags
-        results_text += generate_result_text(query['query_name'], checks_result.checked_output_df)
-
-    flags = pd.Series(flags, dtype=str).drop_duplicates().to_list()
-    if len(flags) == 0:
-        flags.append("GREEN")
-    flag_prefix = f"[{', '.join(flags)}]"
+    with db_manager:
+        results_text = ""
+        flags = []
+        for query_key in query_keys:
+            query = QUERIES[query_key]
+            query_output_df = execute_query_and_write_to_csv(query, db_manager)
+            checks_result = run_checks_on_output(query['checks'], query_output_df)
+            flags += checks_result.flags
+            results_text += generate_result_text(query['query_name'], checks_result.checked_output_df)
+
+    flag_set = set(flags)
+    if len(flag_set) == 0:
+        flag_set.add("GREEN")
+    flag_prefix = f"[{', '.join(flag_set)}]"
     now = datetime.now()
     print(f"{flag_prefix} {job_name} for {now:%B %d, %Y}\n{results_text}")
 
-    # Close connections to data sources
-    manager.tear_down()
-
-    if "RED" in flags:
-        logger.error("Status is RED")
\ No newline at end of file
+    if "RED" in flag_set:
```

Comment:
It doesn't look like we were ever printing anything when the status was Yellow? I don't remember if we ever fully configured the yellow (unstable) status for this job either on Jenkins? I feel like we need to return a specific exit code and have Jenkins check for that? Or maybe it does work?

I don't see [anything that's mentioned here in the config](https://stackoverflow.com/questions/8148122/how-to-mark-a-build-unstable-in-jenkins-when-running-shell-scripts)? 

It might still be worth printing out a final status for yellow at the end here.

Reply:
The results get sent out everyday by email, and if it is yellow, that shows up in the subject line. I found at least one example of that looking in my email.

I think the reason we are logging RED is so that ERROR shows up, which I think is what Jenkins is looking for. We could log YELLOW as a warning, but until we know Jenkins can consume that in a meaningful way for us, I'd say we could just leave this as is?

Reply:
I don't think there's any harm in logging `Status is YELLOW` at the warning level, if you'd like to have that for testing purposes. Let me know.

Reply:
I added a log message at warning level

Reply:
Okay cool, yeah there's no harm to it and I think it Jenkins has to be able to parse the output to look for a pattern to be able to set the status. So this would be a good test when I get to trying that out. 

Though I don't know how we'd actually force it to be Yellow status for testing this works. 🤔